File: 2020/Day15.java
public class Day15 {

    public static void main(String[] args) {
        final int[] challengeInput = {14, 8, 16, 0, 1, 17};
        play(challengeInput, 2020);        // part 1: 240
        play(challengeInput, 300_000_00);  // part 2: 505
    }

    public static void play(final int[] startNumbers, final int endTurn) {
        final int[] memory = new int[endTurn];
        int turn = 0;
        int lastSeen = 0;
        for (int sn : startNumbers) {
            lastSeen = memory[sn];
            memory[sn] = ++turn;
        }
        int num = 0;
        while (turn < endTurn) {
            num = lastSeen == 0 ? 0 : turn - lastSeen;
            lastSeen = memory[num];
            memory[num] = ++turn;
        }
        System.out.println(num);
    }
}

File: 2020/day_15.R
# https://adventofcode.com/2020/day/15


play <- function(numbers, end_turn) {
    # last_num is always +1 because in R vector indices start at 1.
    memory = integer(end_turn)
    for (turn in seq_along(numbers)) {
        last_num = numbers[turn] + 1L
        last_seen = memory[last_num]
        memory[last_num] = turn
    }
    for (turn in seq(turn + 1, end_turn)) {
        if (last_seen > 0) {
            last_num = turn - last_seen
        } else {
            last_num = 1L
        }
        last_seen = memory[last_num]
        memory[last_num] = turn
    }
    last_num - 1L
}


challenge_input = as.numeric(strsplit(readLines("2020/day_15_input.txt"), ",")[[1]])
print(play(challenge_input, 2020))  # 240
print(play(challenge_input, 30000000))  # 505
